### ISSUE TITLE
pyright type inference tripped up by @contextlib.contextmanager

### ISSUE THREAD

**Issue by allisonkarlitskaya** [microsoft/pyright#7386]
```
$ pyright --version
pyright 1.1.352
```

A bit of background: we use a lot of JSON in our project and we have some helpers for type-safe access to it.  They're based around looking up an attribute in a dictionary and enforcing that it's present, with the correct type.  Most of the helpers also allow a "default" value which can be returned if the key is not present.

Our way of getting the types right for this is a bit complicated, but it works properly with both mypy and pyright.  A recent attempted feature addition is causing issues with pyright.  See the (dramatically simplified) example below.

**Describe the bug**

The general idea in the following code is that if we provide a default value then its type must be considered to be a possible return value of the function.  If no value is provided then `DT` is (probably?) inferred down to its most specific possible type: `Never` and the only possible return type is expected type.

```python
import contextlib
from enum import Enum
from typing import Iterator, TypeVar, reveal_type

DT = TypeVar('DT')


# We can't use None as a sentinel because it's often the actual default value
# EllipsisType is difficult because it's not available before 3.10.
# See https://peps.python.org/pep-0484/#support-for-singleton-types-in-unions
class _Empty(Enum):
    TOKEN = 0


_empty = _Empty.TOKEN


def generate_str(default: DT | _Empty = _empty) -> Iterator[DT | str]:
    yield 'foo'


@contextlib.contextmanager
def context_str(default: DT | _Empty = _empty) -> Iterator[DT | str]:
    yield 'foo'


# pyright and mypy correctly infer DT down to Never and resolve the union to 'str'
reveal_type(next(generate_str()))  # should be: str

# mypy gets "str" for this, but pyright falls over with ""DT@context_str | str"
reveal_type(context_str().__enter__())

with context_str() as y:
    # same result as above, pyright says "DT@context_str | str"
    reveal_type(y)  # should be: str
```

pyright's output:

```
narrow.py:28:13 - information: Type of "next(generate_str())" is "str"
narrow.py:31:13 - information: Type of "context_str().__enter__()" is "DT@context_str | str"
narrow.py:35:17 - information: Type of "y" is "DT@context_str | str"
0 errors, 0 warnings, 3 informations 
```

mypy's output:
```
narrow.py:28: note: Revealed type is "builtins.str"
narrow.py:31: note: Revealed type is "builtins.str"
narrow.py:35: note: Revealed type is "builtins.str"
Success: no issues found in 1 source file
```

so: pyright gets things right with the iterator, but wrapping it in `@contextlib.contextmanager` seems to break things.

**Comment by allisonkarlitskaya** [comment]
I went into the typeshed and pulled out the annotations for `@contextlib.contextmanager` and kept working it down until I got a more minimal example:

```python
from typing import Callable, ParamSpec, TypeVar, reveal_type

T = TypeVar("T")
P = ParamSpec("P")


def func_identity(func: Callable[P, T]) -> Callable[P, T]:
    return func


def identity(func: T) -> T:
    return func


def str_or_default(default: T | None = None) -> T | str:
    return 'foo'


# pyright and mypy both say "str"
reveal_type(identity(str_or_default)())

# mypy gets "str" for this, but pyright falls over with "T(1)@str_or_default | str"
reveal_type(func_identity(str_or_default)())
```

I guess this is another bug to do with `ParamSpec` handling...

**Comment by erictraut** [comment]
Thanks, I'm able to repro the issue. I'll investigate further.

**Comment by allisonkarlitskaya** [comment]
> This is addressed in pyright 1.1.358, which I just published.

Thanks very much!
